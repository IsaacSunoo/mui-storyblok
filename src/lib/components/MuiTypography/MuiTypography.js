import React, { createElement, lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import Storyblok from '../../utils/Storyblok';

const MuiLink = lazy(() => import('./components/MuiLink/MuiLink'));
const MuiIcon = lazy(() => import('../MuiIcon/MuiIcon'));
const MuiIconButtonRedirect = lazy(() => import('../MuiIconButtonRedirect/MuiIconButtonRedirect'));
const MuiIconButtonHref = lazy(() => import('../MuiIconButtonHref/MuiIconButtonHref'));
const MuiIconButtonDownload = lazy(() => import('../MuiIconButtonDownload/MuiIconButtonDownload'));
const MuiIconButtonDialog = lazy(() => import('../MuiIconButtonDialog/MuiIconButtonDialog'));

const MuiButtonRedirect = lazy(() => import('../MuiButtonRedirect/MuiButtonRedirect'));
const MuiButtonHref = lazy(() => import('../MuiButtonHref/MuiButtonHref'));
const MuiButtonDownload = lazy(() => import('../MuiButtonDownload/MuiButtonDownload'));
const MuiButtonDialog = lazy(() => import('../MuiButtonDialog/MuiButtonDialog'));

const MuiTooltip = lazy(() => import('../MuiTooltip/MuiTooltip'));

/**
 * MuiTypography is used in storyblok redirect to react routes
 */

export const MuiTypography = ({
  align,
  color,
  variant,
  height,
  width,
  rootClass,
  content,
}) => {
  const components = {
    MuiLink,
    MuiIcon,
    MuiIconButtonRedirect,
    MuiIconButtonHref,
    MuiIconButtonDownload,
    MuiIconButtonDialog,
    MuiButtonRedirect,
    MuiButtonHref,
    MuiButtonDownload,
    MuiButtonDialog,
    MuiTooltip,
  };

  const defaultStyles = {
    height,
    width,
  };
  const styles = Storyblok.arrayToMuiStyles(rootClass, defaultStyles);

  return (
    <Typography
      className={styles.root}
      variant={variant}
      align={align}
      color={color}
    >
      <Suspense fallback={<div />}>
        {content.map((item, index) => {
          // MuiText is in storyblok and only returns text
          if (item.component === 'MuiText') return item.text;
          return createElement(components[item.component], Object.assign(item, { key: index }));
        })}
      </Suspense>
    </Typography>
  );
};

export default MuiTypography;

MuiTypography.propTypes = {
  /** mui prop: 'inherit' | 'left' | 'center' | 'right' | 'justify' */
  align: PropTypes.string,
  /** 'initial'| 'inherit'| 'primary'| 'secondary'| 'textPrimary'| 'textSecondary'| 'error' */
  color: PropTypes.string,
  /** 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'subtitle1' | 'subtitle2' | 'body1' | 'body2' | 'caption' | 'button' | 'overline' | 'srOnly' | 'inherit' */
  variant: PropTypes.string,
  /** hight of the Typography element */
  height: PropTypes.string,
  /** width of the Typography element */
  width: PropTypes.string,

  /** stroyblok multiselect of css classes */
  rootClass: PropTypes.arrayOf(PropTypes.string),

  content: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,
};

MuiTypography.defaultProps = {
  width: '100%',
  height: '100%',
  align: 'inherit',
  color: 'initial',
  variant: 'body1',
  rootClass: [],
};
