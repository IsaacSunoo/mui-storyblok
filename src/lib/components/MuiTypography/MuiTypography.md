MuiTypography example:

```js noeditor
import MuiTypography from './MuiTypography';
```

```js

const muiText = {
    component: 'MuiText',
    text: 'you can concatenate stings',
}

const icon = {
    component: 'MuiIcon',
    iconName: 'done_all',
};

const iconButtonDownload = {
    component: 'MuiIconButtonDownload',
    herf: "https://miro.medium.com/max/1200/1*mk1-6aYaf_Bes1E3Imhc0A.jpeg",
    fileName: "cool.jpeg",
    icon: [icon],
};

const muiText2 = {
    component: 'MuiText',
    text: 'like this',
}

  const content = [muiText, iconButtonDownload, muiText2];

  <MuiTypography
    content={content}
  />
```