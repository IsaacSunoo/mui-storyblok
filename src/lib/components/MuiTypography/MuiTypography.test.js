import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import MuiTypography from './MuiTypography';

function setup() {
  const props = {
    content: [{
      component: 'MuiText',
      text: 'this is text',
    }, {
      component: 'MuiLink',
      text: 'btn',
      href: 'google.com',
    }, {
      component: 'MuiIconButtonRedirect',
      redirectRoute: 'page-welcom',
      iconButton: [{
        component: 'MuiIconButton',
        icon: [{
          component: 'MuiIcon',
          iconName: 'star',
        }],
      }],
    }, {
      component: 'MuiIcon',
      iconName: 'android',
    }],
  };
  const comp = shallow(<MuiTypography {...props} />);
  return { comp, props };
}

describe('<MuiTypography />', () => {
  it('renders MuiTypography', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();

    const tree = renderer.create((
      <MemoryRouter>
        <MuiTypography {...props} />
      </MemoryRouter>
    ));
    expect(tree).toMatchSnapshot();
  });
});
