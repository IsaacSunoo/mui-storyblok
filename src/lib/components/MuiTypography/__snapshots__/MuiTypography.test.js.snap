// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiTypography /> snapshot 1`] = `
<p
  className="MuiTypography-root makeStyles-root-3 MuiTypography-body1"
>
  <div />
</p>
`;
